[pytorch] CMakeLists changes to enable libtorch for Android

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20000 [pytorch] exclude some caffe2 modules from libtorch mobile build
* **#19762 [pytorch] CMakeLists changes to enable libtorch for Android**

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

Differential Revision: [D15087653](https://our.internmc.facebook.com/intern/diff/D15087653)

ljk53 committed:
[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

dzhulgakov commented on CMakeLists.txt:
```diff
@@ -71,6 +71,7 @@ option(BUILD_CUSTOM_PROTOBUF "Build and use Caffe2's own protobuf under third_pa
 option(BUILD_PYTHON "Build Python binaries" ON)
 option(BUILD_CAFFE2_OPS "Build Caffe2 operators" ON)
 option(BUILD_SHARED_LIBS "Build libcaffe2.so" ON)
+option(BUILD_TORCH_MOBILE "Build libtorch for mobile" OFF)
```
There's BUILD_ATEN_MOBILE flag above - is it relevant to expose it to end user?

orionr replied:
Yup - can we unify them? Also, seems like `BUILD_MOBILE` might be the best name if you're willing to tweak it.

orionr replied:
Please add this to cmake/Summary.cmake as well so it's shown in the CMake output.

ljk53 replied:
BUILD_ATEN_MOBILE was not introduced by me so I can only guess its meaning by reading the code - from what I see looks like it's always set when platform is Android/iOS, so it means "we are building mobile library".

Currently with BUILD_ATEN_MOBILE=True it only builds ATen/Core but not builds actual ATen kernels nor TH/THNN kernels.

What this PR does it to include ATen & TH kernels when a new macro is set to true. So there are a few design decisions we need think about here:

1. Do we want to expose the switch of building ATen/TH kernels to end user?
I guess the answer is yes? - if current caffe2 mobile library is still being used by our users. Always building ATen/TH kernels will regress caffe2 mobile library build size until we provide a solution to modularize ATen/TH kernels.

2. How do we structure the build macros?

A - doesn't change BUILD_ATEN_MOBILE meaning, introducing BUILD_TORCH_MOBILE to enable ATen/TH kernels (what this PR does):
```
+-----------------+-------------------+--------------------+
|                 | BUILD_ATEN_MOBILE | BUILD_TORCH_MOBILE |
+-----------------+-------------------+--------------------+
| Mobile platform | ON                |                    |
| Build caffe2go  | ON                | OFF                |
| Build libtorch  | ON                | ON                 |
+-----------------+-------------------+--------------------+
```

B - replace BUILD_ATEN_MOBILE with BUILD_MOBILE, which still enabled for all mobile builds; use BUILD_ATEN (instead of BUILD_TORCH_MOBILE) to build ATen/TH kernels:
```
+-----------------+--------------+------------+-------------+
|                 | BUILD_MOBILE | BUILD_ATEN | BUILD_TORCH |
+-----------------+--------------+------------+-------------+
| Mobile platform | ON           |            |             |
| Build caffe2go  | ON           | OFF        | OFF         |
| Build libtorch  | ON           | ON         | ON          |
+-----------------+--------------+------------+-------------+
```
Note that BUILD_TORCH is an existing build macro - we don't need change it for now. Maybe add BUILD_JIT or something else later to provide fine-grain control for libtorch mobile build.

How does 2B sounds to you?
@kostmo @dzhulgakov @orionr

orionr replied:
2B sounds reasonable to me! Thanks @ljk53.

ljk53 replied:
Updated the PR according to 2B with a few modifications:
- rename BUILD_ATEN_MOBILE to INTERN_BUILD_MOBILE and hide it from users;
- use INTERN_BUILD_ATEN_OPS to control whether build aten ops and set it according to INTERN_BUILD_MOBILE & BUILD_TORCH_MOBILE;
- keep BUILD_TORCH_MOBILE to differentiates caffe2 mobile build and libtorch mobile build;
- How we are going to use BUILD_TORCH_MOBILE:
1. as a master switch BUILD_TORCH_MOBILE automatically enables/disables other macros, e.g.: BUILD_TORCH / BUILD_CAFFE2_OPS;
2. strip out some ATen/TH/Caffe2 code that are not applicable for mobile - directly using INTERN_BUILD_MOBILE for this purpose will affect caffe2 mobile build, e.g. removing threadpool on this PR;

orionr replied:
Eventually, I think Caffe2 mobile builds shouldn't exist. Given that, I'm still tempted to call it BUILD_MOBILE. We can chat offline, though.

ljk53 replied:
Chatted with @orionr offline - we decided to flip and rename BUILD_TORCH_MOBILE to BUILD_CAFFE2_MOBILE and mark it as deprecating. Eventually libtorch will be default mobile build. Keep the BUILD_CAFFE2_MOBILE flag just to preserve build_android.sh functionality for now.

dzhulgakov commented on CMakeLists.txt:
```diff
@@ -167,6 +168,9 @@ endif(MSVC)
 if (ANDROID OR IOS)
   set(BUILD_ATEN_MOBILE ON)
+  if (BUILD_TORCH_MOBILE)
+    set(TORCH_MOBILE ON)
```
you set this var - do you use it anywhere?

ljk53 replied:
@dzhulgakov it is the source code level control to not risk regressing internal facebook product, see discussion: https://fburl.com/hyk9n7c2
It's used by stacked diff(PR).

dzhulgakov commented on cmake/TorchConfig.cmake.in:
```diff
@@ -40,11 +40,19 @@ if (@BUILD_SHARED_LIBS@)
+if (NOT ANDROID)
+  find_library(TORCH_LIBRARY torch PATHS "${TORCH_INSTALL_PREFIX}/lib")
+else()
+  find_library(TORCH_LIBRARY NO_CMAKE_FIND_ROOT_PATH torch PATHS "${TORCH_INSTALL_PREFIX}/lib")
```
just curious - what is this?

orionr replied:
Same. :)

ljk53 replied:
This file is used to generate cmake for library users - I guess @t-vi found it necessary when building his demo app? I can remove them from this PR and do it separately if it's a must have.

orionr replied:
I'm okay with the change if it looks to be required by a test app.

pjh5 replied:
This allows client libraries to add a "find_package(Torch)" line to their CMakeLists.txt without having to know how our libraries/headers/etc. are structured. Ideally I think we should have a separate config for mobile builds though; I can imagine that users that want to build an android package probably also have a server installation of torch that they don't want to interfere with. But that's for a follow-up PR

dzhulgakov reviewed (commented):
I wonder whether it'd be easier to understand the details if we had separate vars for platform (e.g. what current BUILD_ATEN_MOBILE is) and have derived var BUILD_ATEN or BUILD_JIT determined based on it. That way we'd have mostly single conditionals around instead of BUILD_ATEN_MOBILE AND NOT BUILD_TORCH_MOBILE. But it's a soft suggestion, not insisting

orionr commented on aten/CMakeLists.txt:
```diff
@@ -1,14 +1,17 @@
-if (BUILD_ATEN_MOBILE)
+if (BUILD_ATEN_MOBILE AND NOT BUILD_TORCH_MOBILE)
```
With your changes `BUILD_ATEN_MOBILE` should just get replaced with `BUILD_MOBILE` (aka BUILD_TORCH_MOBILE)

ljk53 replied:
added INTERN_BUILD_ATEN_OPS for this condition

orionr commented on caffe2/CMakeLists.txt:
```diff
@@ -4,7 +4,7 @@ include(../cmake/Codegen.cmake)
 
 # ---[ ATen build
-if (NOT BUILD_ATEN_MOBILE)
+if (NOT BUILD_ATEN_MOBILE OR BUILD_TORCH_MOBILE)
```
Let's combine these to `BUILD_MOBILE`

ljk53 replied:
Combined to INTERN_BUILD_ATEN_OPS

orionr commented on cmake/Codegen.cmake:
```diff
@@ -43,7 +43,7 @@ configure_file(
         DESTINATION include
         FILES_MATCHING PATTERN "*.h")
-if (BUILD_ATEN_MOBILE)
+if (BUILD_ATEN_MOBILE AND NOT BUILD_TORCH_MOBILE)
```
Can we just remove this completely? Seems like `BUILD_MOBILE` will  now just do the standard thing. No need for ATen core.

ljk53 replied:
Probably still need this condition as we keep distinction between caffe2 mobile and libtorch mobile. Probably should combine this one into "NOT INTERN_BUILD_ATEN_OPS" as well? if it's not building aten ops, then only install aten/core; otherwise doesn't matter?

orionr commented on cmake/Codegen.cmake:
```diff
@@ -52,7 +52,7 @@ install(FILES ${CMAKE_BINARY_DIR}/caffe2/core/macros.h
 
 # ---[ ATen specific
-if (NOT BUILD_ATEN_MOBILE)
+if (NOT BUILD_ATEN_MOBILE OR BUILD_TORCH_MOBILE)
```
Same here.

ljk53 replied:
Combined into INTERN_BUILD_ATEN_OPS

orionr commented on cmake/Dependencies.cmake:
```diff
@@ -85,6 +85,8 @@ if(NOT BUILD_ATEN_MOBILE)
   set(BLAS "MKL" CACHE STRING "Selected BLAS library")
 else()
   set(BLAS "Eigen" CACHE STRING "Selected BLAS library")
+  set(AT_MKLDNN_ENABLED 0)
```
Shouldn't these be ignored for BLAS "Eigen"? If not, that's probably the better fix to put somewhere.

ljk53 replied:
I think you are right - we shouldn't need explicit set these to 0. The places that set these to 1 below are already covered by "NOT BUILD_ATEN_MOBILE" condition.
@t-vi - any particular reason to set them here?

ljk53 replied:
@orionr looks like these are needed by Config.h.in - we need set some default value otherwise it will cause compile error. Config.h is used by Aten code which was not compiled for mobile (BUILD_ATEN_MOBILE) previously. Now we need set them.

t-vi replied:
I'm a lazy person, so most of the things I put in were needed at some point of time. :slightly_smiling_face:

orionr replied:
Sounds reasonable. This could use some cleanup later, but good for now.

orionr reviewed (commented):
Exciting stuff! Biggest comment is that we should be able to go with just `BUILD_MOBILE` and remove the separate `BUILD_TORCH_MOBILE` and `BUILD_ATEN_MOBILE` options. I think @smessmer and @ezyang might have comments on whether ATen core will be needed then or not.

t-vi commented:
The TorchConfig change is needed to keep CMake from finding host libs when building C++ extensions / custom ops for Android.

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

dzhulgakov reviewed (commented):
The new setup looks good. I'll let other folks take a final look and accept

pjh5 commented on .jenkins/pytorch/build.sh:
```diff
@@ -65,10 +65,10 @@ fi
+  build_args+=("-DBUILD_TORCH_MOBILE=ON")
+  build_args+=("-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')")
+  build_args+=("-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')")
+  build_args+=("-DPYTHON_INCLUDE_DIR=$(python -c 'from distutils.sysconfig import get_python_inc; print(get_python_inc())')")
```
Why is this needed? Does the existing python logic in https://github.com/pytorch/pytorch/blob/master/cmake/Dependencies.cmake#L589 not work?

EDIT: Okay I see that BUILD_PYTHON is turned OFF below so that logic never runs. Why is python needed for the android build in the first place then?

ljk53 replied:
@pjh5 I think PYTHON_EXECUTABLE is needed by aten codegen script; PYTHON_INCLUDE_DIR might not be necessary, let me test it.

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

pjh5 commented on cmake/Summary.cmake:
```diff
@@ -20,6 +20,7 @@ function (caffe2_print_configuration_summary)
 
   message(STATUS "  TORCH_VERSION         : ${TORCH_VERSION}")
   message(STATUS "  CAFFE2_VERSION        : ${CAFFE2_VERSION}")
+  message(STATUS "  BUILD_CAFFE2_MOBILE   : ${BUILD_CAFFE2_MOBILE}")
```
Can you add INTERN_BUILD_MOBILE too?

ljk53 replied:
added in new revision

ljk53 committed:
Update on "[pytorch] CMakeLists changes to enable libtorch for Android"

[pytorch] CMakeLists changes to enable libtorch for Android

Summary:
Stole from https://github.com/pytorch/pytorch/pull/16242
Only include CMakeLists changes to make it easer for reviewers.

Change pytorch android CI job to build libtorch instead of caffe2 -
turns out it's as simple as adding a few cmake macros. Will properly
update tools/build_pytorch_libs.py in subsequent diffs.

Test Plan:
- Make sure pytorch android CI job succeeds;
- Test libtorch.so on Android device;

gh-metadata: pytorch pytorch 19762 gh/ljk53/14/head

t-vi commented:
I'm building PyTorch for Android now. Thank you @ljk53  and everyone involved. You are awesome!

ljk53 commented:
@t-vi thanks for initiating this effort! It's all based on your work :)

kindlychung commented:
Is there a documentation on how to use this on android? Thanks.

ezyang commented:
Check the CI scripts for our Android build; they're the most canonical
reference at the moment.  Folder is  .circleci/pytorch

also cc @ljk53

Edward

Excerpts from Kaiyin Zhong's message of 2019-07-03 06:40:17 -0700:
> Is there a documentation on how to use this on android? Thanks. 
>

ljk53 commented:
What @ezyang said, you can build the current version with scripts/build_android.sh + "BUILD_CAFFE2_MOBILE=OFF" and pytorch environment for codegen:

```
./scripts/build_android.sh \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

The work is not done - we are working on improvements in terms of build size, perf and documentation. Stay tuned!
